juju/model.py

```diff
@@ -132,13 +158,30 @@ def units(self):
     @property
     def data(self):
         if self._data is None:
-            units = [self._local_unit] + self.units
-            # TODO: Restore the RelationData wrapping class. This will prevent unintended side-effects if the
-            # charm code tries to do something unexpected with the return value of this property.
-            self._data = {unit: RelationUnitData(self.relation_id, unit, self._backend) for unit in units}
+            self._data = RelationData(self.relation_name, self.relation_id, self._local_unit, self.units, self._backend)
```

Comment:
I'm not sure all of the indirection here is worth it, rather than just having self.data as the real thing and initializing it with the instance. We're saving the realtion_list call, but assuming we don't actually create the Relation itself until it's looked into via the relation name, there's not much useful in this value that doesn't involve looking into the unit list right now.

Anyway, we can keep it as it is and see if the trade-off pays off at some point. If it doesn't it's easy to make it simpler eventually.

Reply:
The benefit of this indirection here is just to prevent the resulting dictionary from being mutable.  This was done in response to your [comment](https://github.com/canonical/charmbase/pull/14#discussion_r335071559) on PR#14.

Reply:
The comment was about having self._data and descriptors and if branches and etc ... instead of simply self.data.

Reply:
Oooh, I see now.  Yes, +1.